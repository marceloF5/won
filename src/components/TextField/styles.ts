import styled, { css, DefaultTheme } from 'styled-components'
import { TextFieldProps } from '.'

type IconPosition = Pick<TextFieldProps, 'iconPosition'>

type IconProps = {
    hasIcon: boolean
} & IconPosition

const wrapperModifiers = {
    error: (theme: DefaultTheme) => css`
        ${InputWrapper} {
            border-color: ${theme.colors.red};
        }
        ${Label},
        ${Icon} {
            color: ${theme.colors.red};
        }
    `,
    disabled: (theme: DefaultTheme) => css`
        ${Label},
        ${Input},
        ${Icon} {
            color: ${theme.colors.gray};
            cursor: not-allowed;

            &::placeholder {
                color: currentColor;
            }
        }
    `
}

export const Wrapper = styled.div<
    Pick<TextFieldProps, 'disabled'> & { error?: boolean }
>`
    ${({ theme, disabled, error }) => css`
        ${!!error && wrapperModifiers.error(theme)}
        ${disabled && wrapperModifiers.disabled(theme)}
    `}
`

export const InputWrapper = styled.div`
    ${({ theme }) => css`
        display: flex;
        background: ${theme.colors.lightgray};
        border: 0.2rem solid;
        border-color: ${theme.colors.lightgray};
        border-radius: ${theme.border.radius};
        padding: 0 ${theme.spacings.xsmall};

        &:focus-within {
            box-shadow: 0 0 0.5rem ${theme.colors.primary};
        }
    `}
`

export const Input = styled.input<IconProps>`
    ${({ theme, hasIcon, iconPosition }) => css`
        background: transparent;
        border: 0;
        color: ${theme.colors.black};
        font-family: ${theme.font.family};
        font-size: ${theme.font.sizes.medium};
        outline: none;
        padding: ${theme.spacings.xxsmall} 0;
        padding-${iconPosition}: ${hasIcon ? theme.spacings.xxsmall : 0};
        width: 100%;
    `}
`

export const Label = styled.label`
    ${({ theme }) => css`
        font-size: ${theme.font.sizes.small};
        color: ${theme.colors.black};
        cursor: pointer;
    `}
`

export const Icon = styled.div<IconPosition>`
    ${({ theme, iconPosition }) => css`
        display: flex;
        width: 2.2rem;
        color: ${theme.colors.gray};
        order: ${iconPosition === 'right' ? 1 : 0};

        & > svg {
            width: 100%;
        }
    `}
`

export const Error = styled.p`
    ${({ theme }) => css`
        color: ${theme.colors.red};
        margin-top: ${theme.spacings.xxxsmall};
    `}
`
