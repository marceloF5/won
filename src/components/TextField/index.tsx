import { InputHTMLAttributes, useState } from 'react'
import * as S from './styles'

export type TextFieldProps = {
    onInputChange?: (value: string) => void
    disabled?: boolean
    error?: string
    label?: string
    name?: string
    labelColor?: 'white' | 'black'
    icon?: JSX.Element
    iconPosition?: 'left' | 'right'
    initialValue?: string
} & InputHTMLAttributes<HTMLInputElement>

const TextField = ({
    onInputChange,
    disabled = false,
    error = '',
    label,
    name,
    icon,
    iconPosition = 'left',
    initialValue,
    ...otherProps
}: TextFieldProps) => {
    const [value, setValue] = useState(initialValue)

    const handleOnChange = (e: React.ChangeEvent<HTMLInputElement>) => {
        const newValue = e.target.value
        setValue(newValue)

        !!onInputChange && onInputChange(newValue)
    }

    return (
        <S.Wrapper disabled={disabled} error={!!error}>
            {!!label && <S.Label htmlFor={name}>{label}</S.Label>}
            <S.InputWrapper>
                {!!icon && <S.Icon iconPosition={iconPosition}>{icon}</S.Icon>}
                <S.Input
                    type="text"
                    onChange={handleOnChange}
                    value={value}
                    hasIcon={!!icon}
                    iconPosition={iconPosition}
                    disabled={disabled}
                    name={name}
                    {...(label ? { id: name } : {})}
                    {...otherProps}
                ></S.Input>
            </S.InputWrapper>
            {!!error && <S.Error>{error}</S.Error>}
        </S.Wrapper>
    )
}

export default TextField
