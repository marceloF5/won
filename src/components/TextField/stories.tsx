import { Story, Meta } from '@storybook/react/types-6-0'
import { Email } from '@styled-icons/material-outlined'
import TextField, { TextFieldProps } from '.'

export default {
    title: 'atoms/Form/TextField',
    component: TextField,
    args: {
        disabled: false,
        label: 'E-mail',
        name: 'email',
        initialValue: 'mp.fortunato88@gmail.com',
        placeholder: 'E-mail'
    },
    argTypes: {
        onInput: { action: 'changed' },
        icon: { type: '' }
    }
} as Meta

export const Default: Story<TextFieldProps> = (args) => (
    <div style={{ maxWidth: 300, padding: 15 }}>
        <TextField {...args} />
    </div>
)

export const withIcon: Story<TextFieldProps> = (args) => (
    <div style={{ maxWidth: 300, padding: 15 }}>
        <TextField {...args} />
    </div>
)
withIcon.args = {
    icon: <Email />
}

export const withError: Story<TextFieldProps> = (args) => (
    <div style={{ maxWidth: 300, padding: 15 }}>
        <TextField {...args} />
    </div>
)
withError.args = {
    icon: <Email />,
    error: 'Opps... invalid field'
}
