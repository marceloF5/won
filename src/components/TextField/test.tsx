import { screen, waitFor } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import { Email } from '@styled-icons/material-outlined'
import { renderWithTheme } from 'utils/tests/helpers'

import TextField from '.'

describe('<TextField />', () => {
    it('should render with label', () => {
        renderWithTheme(<TextField label="TextField" name="LabelName" />)

        expect(screen.getByLabelText('TextField')).toBeInTheDocument()
    })

    it('should render without label', () => {
        renderWithTheme(<TextField />)

        expect(screen.queryByLabelText('TextField')).not.toBeInTheDocument()
    })

    it('should render with placeholder', () => {
        renderWithTheme(<TextField placeholder="hey you" />)

        expect(screen.getByPlaceholderText('hey you')).toBeInTheDocument()
    })

    it('should changes its value when typing', async () => {
        const onInput = jest.fn()
        renderWithTheme(
            <TextField onInput={onInput} label="TextField" name="TextField" />
        )

        const input = screen.getByRole('textbox')
        const text = 'This is my text'
        userEvent.type(input, text)

        await waitFor(() => {
            expect(input).toHaveValue(text)
            expect(onInput).toHaveBeenCalledTimes(text.length)
        })

        expect(onInput).toHaveBeenCalledWith(text)
    })

    it('should be accessible by tab', () => {
        renderWithTheme(<TextField label="TextField" name="Textfield" />)

        const input = screen.getByLabelText('TextField')
        expect(document.body).toHaveFocus()

        userEvent.tab()
        expect(input).toHaveFocus()
    })

    it('should render with icon', () => {
        renderWithTheme(<TextField icon={<Email data-testid="icon" />} />)

        expect(screen.getByTestId('icon')).toBeInTheDocument()
    })

    it('should render with icon on the right side', () => {
        renderWithTheme(
            <TextField
                icon={<Email data-testid="icon" />}
                iconPosition="right"
            />
        )

        expect(screen.getByTestId('icon').parentElement).toHaveStyle({
            order: 1
        })
    })

    it('should NOT change value when input is disabled', async () => {
        const onInput = jest.fn()
        renderWithTheme(
            <TextField label="TextField" name="TextField" disabled />
        )

        expect(screen.getByRole('textbox')).toBeDisabled()

        const input = screen.getByRole('textbox')
        const text = 'This is my text'
        userEvent.type(input, text)

        await waitFor(() => {
            expect(input).not.toHaveValue(text)
        })

        expect(onInput).not.toHaveBeenCalled()
    })

    it('should NOT accessible when input is disabled', async () => {
        renderWithTheme(
            <TextField label="TextField" name="TextField" disabled />
        )

        const input = screen.getByLabelText('TextField')
        expect(document.body).toHaveFocus()

        userEvent.tab()
        expect(input).not.toHaveFocus()
    })

    it('should render with error', () => {
        const { container } = renderWithTheme(
            <TextField
                label="TextField"
                name="TextField"
                disabled
                error="Opps... invalid field"
            />
        )

        expect(screen.getByText('Opps... invalid field')).toBeInTheDocument()
        expect(container.firstChild).toMatchSnapshot()
    })
})
