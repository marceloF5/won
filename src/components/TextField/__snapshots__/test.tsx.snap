// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TextField /> should render with error 1`] = `
.c0 .c3 {
  border-color: #ff3347;
}

.c0 .c1,
.c0 .sc-eCssSg {
  color: #ff3347;
}

.c0 .c1,
.c0 .c5,
.c0 .sc-eCssSg {
  color: #8f8f8f;
  cursor: not-allowed;
}

.c0 .c1::-webkit-input-placeholder,
.c0 .c5::-webkit-input-placeholder,
.c0 .sc-eCssSg::-webkit-input-placeholder {
  color: currentColor;
}

.c0 .c1::-moz-placeholder,
.c0 .c5::-moz-placeholder,
.c0 .sc-eCssSg::-moz-placeholder {
  color: currentColor;
}

.c0 .c1:-ms-input-placeholder,
.c0 .c5:-ms-input-placeholder,
.c0 .sc-eCssSg:-ms-input-placeholder {
  color: currentColor;
}

.c0 .c1::placeholder,
.c0 .c5::placeholder,
.c0 .sc-eCssSg::placeholder {
  color: currentColor;
}

.c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  background: #eaeaea;
  border: 0.2rem solid;
  border-color: #eaeaea;
  border-radius: 0.4rem;
  padding: 0 1.6rem;
}

.c4:focus-within {
  box-shadow: 0 0 0.5rem #f231a5;
}

.c6 {
  background: transparent;
  border: 0;
  color: #030517;
  font-family: Poppins,-apple-system,BlinkMacSystemFont,'Segoe UI',Roboto,Oxygen,Ubuntu,Cantarell,'Open Sans','Helvetica Neue',sans-serif;
  font-size: 1.6rem;
  outline: none;
  padding: 0.8rem 0;
  padding-left: 0;
  width: 100%;
}

.c2 {
  font-size: 1.4rem;
  color: #030517;
  cursor: pointer;
}

.c7 {
  color: #ff3347;
  margin-top: 0.4rem;
}

<div
  class="c0"
  disabled=""
>
  <label
    class="c1 c2"
    for="TextField"
  >
    TextField
  </label>
  <div
    class="c3 c4"
  >
    <input
      class="c5 c6"
      disabled=""
      id="TextField"
      name="TextField"
      type="text"
      value=""
    />
  </div>
  <p
    class="c7"
  >
    Opps... invalid field
  </p>
</div>
`
